crates/typst/src/diag.rs

```diff
@@ -298,27 +310,51 @@ pub type HintedStrResult<T> = Result<T, HintedString>;
 /// A string message with hints.
 #[derive(Debug, Clone, Eq, PartialEq, Hash)]
 pub struct HintedString {
+    /// The string's components as a single vector of strings.
+    /// The first element of the vector contains the message.
+    /// The remaining elements are the hints.
+    /// This is done to reduce the size of a HintedString.
+    /// The vector is guaranteed to not be empty.
+    components: EcoVec<EcoString>,
+}
+
+impl HintedString {
+    /// Creates a new hinted string with the given message and hints.
+    pub fn new(message: EcoString, hints: impl IntoIterator<Item = EcoString>) -> Self {
```

Comment:
I think the constructor should not take the hints. Rather, we would replicate the full set of `hint`, `with_hint`, and `with_hints` from `SourceDiagnostic`. That seems like a more convenient API to me.